src/server/parser.rs

```diff
@@ -931,4 +816,185 @@ mod test {
+        io.push_bytes("GET / HTTP/1.1\n\
+            Content-Length: 0\n\
+            Connection: close\n\n"
+                          .as_bytes());
+        let m = Stream::<Parser<Proto, MemIo>>::accepted(io.clone(), &mut lp.scope(1))
+                    .expect_machine();
+        m.ready(EventSet::readable(), &mut lp.scope(1))
+         .expect_machine();
+        assert_eq!(*lp.ctx(),
+                   Context {
+                       progressive: false,
+                       headers_received: 1,
+                       body: String::from(""),
+                       chunks_received: 0,
+                       requests_received: 1,
+                   });
+    }
+
+    // FIXME
+    #[test]
+    #[should_panic]
+    fn test_pipelining() {
+        let mut io = MemIo::new();
+        let mut lp = MockLoop::new(Default::default());
+        io.push_bytes("GET /foo HTTP/1.1\r\n\
+            Host: example.com\r\n\
+            Content-Length: 0\r\n\
+            \r\n\
+            GET /bar HTTP/1.1\r\n\
+            Host: example.com\r\n\
+            Content-Length: 0\r\n\
+            \r\n"
+                          .as_bytes());
+        println!("{:?}", io);
+        let m = Stream::<Parser<Proto, MemIo>>::accepted(io.clone(), &mut lp.scope(1))
+                    .expect_machine();
+        m.ready(EventSet::readable(), &mut lp.scope(1))
+         .expect_machine();
+        assert_eq!(*lp.ctx(),
+                   Context {
+                       progressive: false,
+                       headers_received: 2,
+                       body: String::from(""),
+                       chunks_received: 0,
+                       requests_received: 2,
+                   });
+    }
+
+    #[test]
+    fn test_leading_whitespace() {
+        let mut io = MemIo::new();
+        let mut lp = MockLoop::new(Default::default());
+        io.push_bytes("\r\nGET /foo HTTP/1.1\r\n\
+            Host: example.com\r\n\r\n"
+                          .as_bytes());
+        let m = Stream::<Parser<Proto, MemIo>>::accepted(io.clone(), &mut lp.scope(1))
+                    .expect_machine();
+        m.ready(EventSet::readable(), &mut lp.scope(1))
+         .expect_machine();
+        assert_eq!(*lp.ctx(),
+                   Context {
+                       progressive: false,
+                       headers_received: 1,
+                       body: String::from(""),
+                       chunks_received: 0,
+                       requests_received: 1,
+                   });
+    }
+
+    #[test]
+    fn test_broken_http() {
+        let mut io = MemIo::new();
+        let mut lp = MockLoop::new(Default::default());
+        io.push_bytes("GET / HTTP/1.1\r\n\nHost: host\r\n\r\n".as_bytes());
+        let m = Stream::<Parser<Proto, MemIo>>::accepted(io.clone(), &mut lp.scope(1))
+                    .expect_machine();
+        m.ready(EventSet::readable(), &mut lp.scope(1))
+         .expect_machine();
```

Comment:
This test is wrong. "expect_machine" both times works, because  it stops on the first header ("GET / HTTP/1.1\r\n\n") and doesn't even parse the second one (because state machine never sends a reply).